private/bufpkg/bufimage/bufimageutil/image_index.go

```diff
@@ -102,11 +114,27 @@ func newImageIndexForImage(image bufimage.Image, opts *imageFilterOptions) (*ima
 				}
 			}
 
-			index.ByName[string(name)] = descriptor
-			index.ByDescriptor[descriptor] = elementInfo{
-				fullName: string(name),
-				parent:   parent,
-				file:     fileName,
+			// certain descriptor types don't need to be indexed:
+			//  enum values, normal (non-extension) fields, and oneofs
+			var includeInIndex bool
+			switch d := descriptor.(type) {
+			case *descriptorpb.EnumValueDescriptorProto, *descriptorpb.OneofDescriptorProto:
+				// do not add to package elements; these elements are implicitly included by their enclosing type
+			case *descriptorpb.FieldDescriptorProto:
+				// only add to elements if an extension (regular fields implicitly included by containing message)
+				includeInIndex = d.Extendee != nil
+			default:
+				includeInIndex = true
+			}
+
+			if includeInIndex {
```

Comment:
Because of this, type filtering is actually now _faster_ than before. While there is some small overhead of indexing the packages, it is dwarfed by the improvement of no longer including fields and enum values in the index (the most numerous elements in a typical schema). So we do a lot less map insertions and map growing/re-allocating. The fields, oneofs, and enum values aren't actually useful in the index anyway since they can't be referenced by other elements, are not allowed elements to name in the include list, and are never filtered (an enum always includes all of its values; a message always includes all of its oneofs and regular fields).

Before:
```
BenchmarkFilterImage-10    	   28686	     41532 ns/op	   50227 B/op	     186 allocs/op
```
After:
```
BenchmarkFilterImage-10    	   45050	     26201 ns/op	   20824 B/op	     196 allocs/op
```